<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [3]:
# текст запроса
query_3_1 = f'''select 
               count (*) as cnt
               from vacancies 
            '''

In [4]:
# результат запроса
count_vacatiees = pd.read_sql_query(query_3_1, connection)
count_vacatiees

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\295035935.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_vacatiees = pd.read_sql_query(query_3_1, connection)


cnt
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [5]:
# текст запроса
query_3_2 = f'''select
               count (*) as cnt
               from employers 
            '''

In [6]:
# результат запроса
count_employers = pd.read_sql_query(query_3_2, connection)
count_employers

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\2124197464.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_employers = pd.read_sql_query(query_3_2, connection)


cnt
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [7]:
# текст запроса
query_3_3 = f'''select 
               count (*) as cnt
               from areas 
            '''

In [8]:
# результат запроса
count_areas= pd.read_sql_query(query_3_3, connection)
count_areas

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\576246568.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_areas= pd.read_sql_query(query_3_3, connection)


cnt
0  1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [9]:
# текст запроса
query_3_4 = f'''select 
            count (*) as cnt
            from industries 
         '''

In [10]:
# результат запроса
count_industries= pd.read_sql_query(query_3_4, connection)
count_industries

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\2062374590.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_industries= pd.read_sql_query(query_3_4, connection)


cnt
0  294

***

# Выводы по предварительному анализу данных
С помощью реализованныз запросов выявили количество вакансий, работодателей, регионов и сфер деятельности. Результаты получены и соответствуют контрольным данным,
а значит реализация запросов выполнена корректно. 


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [11]:
# текст запроса
query_4_1 = f'''select 
                ar.name as name,
                count (vc.id) as cnt
                from vacancies as vc
                join areas as ar on vc.area_id = ar.id
                group by ar.id
                order by count (vc.id) desc
                limit 5
         '''

In [12]:
# результат запроса
count_vacatiees_in_areas= pd.read_sql_query(query_4_1, connection)
count_vacatiees_in_areas

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\590529597.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_vacatiees_in_areas= pd.read_sql_query(query_4_1, connection)


name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [13]:
# текст запроса
query_4_2 = f'''select 
                count (vc.id) as cnt
                from vacancies as vc
                where (vc.salary_from is not null) or (vc.salary_to is not null)
         '''

In [14]:
# результат запроса
count_vacatiees_with_salary= pd.read_sql_query(query_4_2, connection)
count_vacatiees_with_salary

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\2607444464.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_vacatiees_with_salary= pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [15]:
# текст запроса
query_4_3 = f'''select 
                round(avg (vc.salary_from)) as medium_low_devide,
                round(avg (vc.salary_to)) as medium_high_devide
                from vacancies as vc
                where (vc.salary_from is not null) or (vc.salary_to is not null)
         '''

In [16]:
# результат запроса
count_medium_meaning= pd.read_sql_query(query_4_3, connection)
count_medium_meaning

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\2874259797.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_medium_meaning= pd.read_sql_query(query_4_3, connection)


medium_low_devide  medium_high_devide
0            71065.0            110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [17]:
# текст запроса
query_4_4 = f'''select 
                vc.schedule as schedule,
                vc.employment as employment,
                count (*) as cnt
                from vacancies as vc
                group by schedule, employment
                order by count (*) desc
                offset 1 limit 1
         '''

In [18]:
# результат запроса
count_comdinations_vacaties= pd.read_sql_query(query_4_4, connection)
count_comdinations_vacaties

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\1312073554.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_comdinations_vacaties= pd.read_sql_query(query_4_4, connection)


schedule        employment   cnt
0  Удаленная работа  Полная занятость  7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [19]:
# текст запроса
query_4_5 = f'''select 
                vc.experience as experience
                from vacancies as vc
                group by vc.experience
                order by count (*) 
         '''

In [20]:
# результат запроса
array_experience= pd.read_sql_query(query_4_5, connection)
array_experience

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\3562691138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  array_experience= pd.read_sql_query(query_4_5, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

# Выводы по детальному анализу вакансий
С помощью реализованныз запросов выявили, что вилка средней зарплаты варируется от 71 до 110,5 тыс. руб. Наиболее популярный для поиска работы регион - Москва. 
Вторая по популярности комбинация рабочего графика и типа трудоустройства - удаленная работа - полная занятость. Наибольшее количество вакансий для кондидатов с опытом работы от 1 до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [21]:
# текст запроса
query_5_1 = f'''select 
                emp.name as name,
                count (vc.id) as cnt
                from employers as emp
                join vacancies as vc on vc.employer_id = emp.id
                group by emp.id
                order by count (vc.id) desc
                limit 5
            '''

In [22]:
# результат запроса
popural_employers= pd.read_sql_query(query_5_1, connection)
popural_employers

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\1400030148.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  popural_employers= pd.read_sql_query(query_5_1, connection)


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [23]:
# текст запроса
query_5_2 = f'''select
                ar.name as name,
                count(distinct emp.id) as cnt_emp, 
                count(distinct vc.id) as cnt_vc      
                from areas as ar
                left join employers as emp on ar.id = emp.area
                left join vacancies as vc on ar.id = vc.area_id     
                group by 1
                order by 3, 2 desc
         '''

In [24]:
# результат запроса
area_without_vacations= pd.read_sql_query(query_5_2, connection)
area_without_vacations

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [25]:
#nекст запроса
query_5_3 = f'''select  
                emp.name as name,   
                count (distinct vc.area_id) as cnt
                from employers as emp
                join vacancies as vc on emp.area = vc.area_id
                group by emp.name
                order by cnt desc
         '''

In [26]:
#результат запроса
areas_of_employers= pd.read_sql_query(query_5_3, connection)
areas_of_employers

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [27]:
# текст запроса
query_5_4 = f'''select 
                count (emp.id) as cnt
                from employers as emp
                left join employers_industries as emp_i on emp_i.employer_id = emp.id
                where emp_i.industry_id is null
                order by count (emp.id) desc
            '''

In [28]:
# результат запроса
employers_with_industry= pd.read_sql_query(query_5_4, connection)
employers_with_industry

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\1465365662.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employers_with_industry= pd.read_sql_query(query_5_4, connection)


cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [29]:
# текст запроса
query_5_5 = f'''select
                emp.name as name    
                from employers_industries as emp_i
                left join employers as emp on emp_i.employer_id = emp.id
                left join industries as ind on emp_i.industry_id = ind.id   
                group by emp.name 
                having count(length(ind.name) - length(replace(ind.name, ',', '')) + 1) = 4
                order by emp.name 
                offset 2 limit 1
         '''

In [30]:
# результат запроса
company_name= pd.read_sql_query(query_5_5, connection)
company_name

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\261228651.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  company_name= pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [31]:
# текст запроса
query_5_6 = f'''select
                count(emp.id) as cnt      
                from employers_industries as emp_i
                left join employers as emp on emp_i.employer_id = emp.id
                left join industries as ind on emp_i.industry_id = ind.id    
                where ind.name like 'Разработка программного обеспечения%'
         '''

In [32]:
company_count= pd.read_sql_query(query_5_6, connection)
company_count

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\3836314044.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  company_count= pd.read_sql_query(query_5_6, connection)


cnt
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [33]:
# текст запроса
query_5_7 = f'''select 
                ar.name as name,
   	            count(*) as cnt
                from employers as emp
     	        join vacancies as vc on emp.id = vc.employer_id
     	        join areas as ar on ar.id = vc.area_id
                where emp.name = 'Яндекс'
  	            and ar.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
                 	'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск',
                  	'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар',
                       'Воронеж',  'Пермь', 'Волгоград')
                group by ar.id
                
                union all

                select 
                'Total', 
                count(*)
                from vacancies as vc
 	            join employers as emp on emp.id = vc.employer_id
 	            join areas as ar on ar.id = vc.area_id
                where emp.name = 'Яндекс'
  	            and ar.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
                 	'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск',
                 	'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар',
                       'Воронеж',  'Пермь', 'Волгоград') 
                order by cnt
         '''

In [34]:
count_in_cities= pd.read_sql_query(query_5_7, connection)
count_in_cities

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\112712440.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_in_cities= pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7            Самара   26
8               Уфа   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

# Выводы по анализу работодателей
В рамках анализа выявили, что Яндекс - первая по числу вакансий компания (181 регионов, в 17 городах-милионниках России 485 вакансий.) Среди регионов, в которых нет вакансий, Россия - наибольшая, в которой наибольшее количество работодателей. Количество компаний, в которых не указаны сферы деятельности - 8419. Общее количество компаний, которые так или иначе занимаются программным обеспечением - 3533. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [35]:
# текст запроса
query_6_1 = f'''select 
                count (vc.id) as cnt 
                from vacancies as vc
                where lower(vc.name) like '%data%' 
                or lower(vc.name) like '%данн%' 
         '''

In [36]:
# результат запроса
count_data_vacancies= pd.read_sql_query(query_6_1, connection)
count_data_vacancies

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\932197298.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_data_vacancies= pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [37]:
# текст запроса
query_6_2 = f'''select 
                count (*) as cnt
                from vacancies as vc
                where (vc.name ilike '%data scientist%' 
                or vc.name ilike '%data% cience%' 
                or vc.name ilike '%исследователь данных%' 
                or (vc.name like '%ML%'
                and vc.name not ilike '%HTML%')
                or vc.name ilike '%machine learning%'
                or vc.name ilike '%машинн%обучен%')
                and (vc.name ilike '%junior%'
                or vc.experience ilike '%Нет%опыта%'
                or vc.employment ilike '%Стажировка%')
         '''

In [38]:
# результат запроса
count_junior_vacancies= pd.read_sql_query(query_6_2, connection)
count_junior_vacancies

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\635251388.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_junior_vacancies= pd.read_sql_query(query_6_2, connection)


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [39]:
# текст запроса
query_6_3 = f'''select 
                count (*) as cnt
                from vacancies as vc
                where ( vc.name ilike '%data scientist%'
             	 or vc.name ilike '%data science%'
             	 or vc.name ilike '%исследователь данных%'
                or (vc.name like '%ML%' and vc.name not ilike '%HTML%')
             	 or vc.name ilike '%machine learning%'
                or vc.name ilike '%машинн%обучен%')
             	 and (vc.key_skills ilike '%postgres%'
                or vc.key_skills ilike '%SQL%')
         '''

In [40]:
# результат запроса
vacancies_with_sql= pd.read_sql_query(query_6_3, connection)
vacancies_with_sql

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\3808662879.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_with_sql= pd.read_sql_query(query_6_3, connection)


cnt
0  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [41]:
# текст запроса
query_6_4 = f'''select 
                count (*) as cnt
                from vacancies as vc
                where ( vc.name ilike '%data scientist%'
                or vc.name ilike '%data science%'
             	 or vc.name ilike '%исследователь данных%'
                or (vc.name like '%ML%' and vc.name not ilike '%HTML%')
             	 or vc.name ilike '%machine learning%'
                or vc.name ilike '%машинн%обучен%')
             	 and vc.key_skills ilike '%python%'
         '''

In [42]:
# результат запроса
vacancies_with_python= pd.read_sql_query(query_6_4, connection)
vacancies_with_python

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\4135760756.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vacancies_with_python= pd.read_sql_query(query_6_4, connection)


cnt
0  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [43]:
# текст запроса
query_6_5 = f'''select 
                round(avg(length(vc.key_skills) - length(replace(vc.key_skills, CHR(9), ''))+1), 2) as avg_ks
                from vacancies as vc
                where ( vc.name ilike '%data scientist%'
             	 or vc.name ilike '%data science%'
             	 or vc.name ilike '%исследователь данных%'
                or (vc.name like '%ML%' and vc.name not ilike '%HTML%')
             	 or vc.name ilike '%machine learning%'
                or vc.name ilike '%машинн%обучен%')
         '''

In [44]:
# результат запроса
count_key_skills= pd.read_sql_query(query_6_5, connection)
count_key_skills

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\3314802207.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_key_skills= pd.read_sql_query(query_6_5, connection)


avg_ks
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [45]:
# текст запроса
query_6_6 = f'''select 
                distinct vac.experience as experience,
                round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) as avg_salary
                from vacancies as vac
                where (vac.name ilike '%data scientist%'
             	or vc.name ilike '%data science%'
             	or vc.name ilike '%исследователь данных%'
                or (vc.name like '%ML%' and vc.name not ilike '%HTML%')
             	or vc.name ilike '%machine learning%'
                or vc.name ilike '%машинн%обучен%')
                and ((vc.salary_from is not null or vc.salary_to is not null) 
                and vc.experience is not null)
                group by vc.experience
         '''

In [46]:
# результат запроса
ds_salary= pd.read_sql_query(query_6_6, connection)
ds_salary

C:\Users\akoty\AppData\Local\Temp\ipykernel_10340\1752030221.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  ds_salary= pd.read_sql_query(query_6_6, connection)


DatabaseError: Execution failed on sql 'select 
                distinct vac.experience as experience,
                round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) as avg_salary
                from vacancies as vac
                where (vac.name ilike '%data scientist%'
             	or vc.name ilike '%data science%'
             	or vc.name ilike '%исследователь данных%'
                or (vc.name like '%ML%' and vc.name not ilike '%HTML%')
             	or vc.name ilike '%machine learning%'
                or vc.name ilike '%машинн%обучен%')
                and ((vc.salary_from is not null or vc.salary_to is not null) 
                and vc.experience is not null)
                group by vc.experience
         ': missing FROM-clause entry for table "vc"
LINE 6:               or vc.name ilike '%data science%'
                         ^


***

# Выводы по предметному анализу
Среди всех вакансий 1771 так или иначе относятся к данным. Среди них 51 вакансия подойдет для начинающего специалиста в дата сайентс. Из общего числа для 201 вакансии указано о необходимости знания SQL, а для 351 - знание Python. В среднем для DS указывают 6 ключевых навыков. Средние зарплаты специалистов варируются от 74.5 до 243 тыс.руб. в зависимости от опыта работы.

# Общий вывод по проекту

# Подведем итог исследования:
1) Проанализировав рынок вакансий видно, что лидирующая область - IT.
2) Из чуть более 49 тыс вакансий почти 2 тыс. относятся к данным и работе с ними, что являтеся не плохим показателем востребованности. 
3) При средней нижней границы заработной платы в 71 тыс.руб., средняя зарплата специалиста DS без опыта работы составляет 74.5 тыс.руб, что выше рынка.
4) Большое количество вокансий в крупных городах, при этом один из самых популярных типов работчих графиков - удаленная работа.
5) Высокая ценность на рынке труда для специалистов DS накладывает обязательства по владению большого количества навыков. 